src/media/js/apps_buttons.js

```diff
@@ -327,17 +327,29 @@ define('apps_buttons',
     }
 
     function markBtnsAsInstalled() {
-        /* For each installed app, look for respective buttons and mark as
-           ready to launch ("Open"). */
+        /*
+            For each install button, check if its manifest URL is installed
+            according to mozApps. If so, mark it as installed to say Open.
+            This is more scalable than looping through all installed apps,
+            querying the DOM for each manifest URL, and checking if its
+            installed.
+        */
         if (!caps.webApps) {
             return;
         }
         apps.getInstalled().done(function(installedApps) {
+            // Create a map to get O(1) lookup.
+            var installedAppsMap = {};
```

Comment:
The installers do this kind of thing fairly often, they have a `getApp()` private method that just loops over the list everytime it's called to find the right app instance. Maybe we should refactor this to expose a map of app instances (manifest URLs would be the keys, as in your code) and then use that everywhere in the installers and here.